src/neuroconv/tools/roiextractors/roiextractors.py

```diff
@@ -672,15 +680,56 @@ def image_mask_iterator():
                     description="1 if ROI was rejected or 0 if accepted as a cell during segmentation operation",
                 ),
             ],
-            id=roi_ids,
         )
+        if mask_type != "pixel" and mask_type != "voxel":  # roi_ids will not be written row-wise
+            plane_segmentation_kwargs.update(id=roi_ids)
+
         plane_segmentation = PlaneSegmentation(**plane_segmentation_kwargs)
+
         if include_roi_centroids:
             tranpose_roi_location_axes = (1, 0) if len(segmentation_extractor.get_image_size()) == 2 else (1, 0, 2)
             roi_locations = segmentation_extractor.get_roi_locations()[tranpose_roi_location_axes, :].T
             plane_segmentation.add_column(
                 name="ROICentroids", description="The x, y, (z) centroids of each ROI.", data=roi_locations
             )
+
+        if mask_type is not "None" and mask_type == "image":
+            plane_segmentation.add_column(
+                name="image_mask",
+                description="Image masks for each ROI.",
+                data=H5DataIO(DataChunkIterator(image_mask_iterator(), **iterator_options), **compression_options),
+            )
+        elif mask_type is not "None" and (mask_type == "pixel" or mask_type == "voxel"):
+            pixel_masks = segmentation_extractor.get_roi_pixel_masks()
+            num_pixel_dims = pixel_masks[0].shape[1]
+
+            assert num_pixel_dims in [3, 4], (
+                "The segmentation extractor returned a pixel mask that is not 3- or 4- dimensional! "
+                "Please open a ticket with https://github.com/catalystneuro/roiextractors/issues"
+            )
+            if mask_type == "pixel" and num_pixel_dims == 4:
+                warn(
+                    "Specified mask_type='pixel', but ROIExtractors returned 4-dimensional masks. "
+                    "Using mask_type='voxel' instead."
+                )
+                mask_type = "voxel"
+            if mask_type == "voxel" and num_pixel_dims == 3:
+                warn(
+                    "Specified mask_type='voxel', but ROIExtractors returned 3-dimensional masks. "
+                    "Using mask_type='pixel' instead."
+                )
+                mask_type = "pixel"
+
+            mask_type_kwarg = f"{mask_type}_mask"
+            for roi_id, pixel_mask in zip(roi_ids, pixel_masks):
+                plane_segmentation.add_roi(**{"id": roi_id, mask_type_kwarg: [tuple(x) for x in pixel_mask]})
+            # plane_segmentation.add_column(
+            #     name=mask_type_kwarg,
+            #     description=f"{mask_type}.capitalize() masks for each ROI.",
+            #     data=[[tuple(x) for x in pixel_mask] for pixel_mask in pixel_masks],
+            #     index=2,
+            # )
```

Comment:
This part was the biggest difficulty - I would prefer to write it column-wise to allow compression, but the indexed (indeed, doubly jagged) nature of it did not play well with the standard ways of adding it as a column.

This may be an unintended consequence of the NWB schema definition of these types, which is not in terms of a 'dataset'  with 'shapes' but as individual x/y/z properties? https://nwb-schema.readthedocs.io/en/latest/format.html#sec-planesegmentation-src

Reply:
Yes, I see, alt ought their type is still `VectorData` so maybe there is a way.

Reply:
Well, they have to be VectorData above because you're passing it directly as the `columns` of the parent `DynamicTable` of the `PlaneSegmentation`, which is separate from calling `add_column` on it.

Critically, this needs `index=2`, but VectorData does not have that: https://hdmf.readthedocs.io/en/stable/hdmf.common.table.html#hdmf.common.table.VectorData (since the meaning of `index` is actually to pair a flat VectorData with a VectorIndex)

Reply:
Just gonna splinter this into #157